photon-client/src/main/scala/com/linkedin/photon/ml/Driver.scala

```diff
@@ -344,58 +353,60 @@ protected[ml] class Driver(
-
-        val tracker = optimizationStatesTrackerOpt.get
-        val perIterationMetrics = tracker
-          .getTrackedStates
-          .map { optimizerState =>
-            val model = glm.updateCoefficients(Coefficients(optimizerState.coefficients, None))
-            val metrics = Evaluation.evaluate(model, validationData)
-
-            metrics
-              .keys
-              .toSeq
-              .sorted
-              .foreach { metric =>
-                logger.info(f"Iteration: [${optimizerState.iter}%6d] Metric: [$metric] value: ${metrics(metric)}")
-              }
-
-            (model, metrics)
-          }
-
-          val finalMetrics = perIterationMetrics.last._2
-
-          perModelMetrics += (regularizationWeight -> finalMetrics)
-          sendEvent(
-            PhotonOptimizationLogEvent(
-              regularizationWeight,
-              optimizationStatesTrackerOpt,
-              Some(perIterationMetrics),
-              Some(finalMetrics)))
+      lambdaModelAndTrackers.foreach {
+        case (regularizationWeight: Double, glm: GeneralizedLinearModel, optimizationStatesTracker: OptimizationStatesTracker) =>
+
+          logger.info(s"Model with lambda = $regularizationWeight:")
+
+          val perIterationMetrics = optimizationStatesTracker
+            .getTrackedStates
+            .map { optimizerState =>
+              val model = glm.updateCoefficients(Coefficients(optimizerState.coefficients, None))
+              val metrics = Evaluation.evaluate(model, validationData)
+
+              metrics
+                .keys
+                .toSeq
+                .sorted
+                .foreach { metric =>
+                  logger.info(f"Iteration: [${optimizerState.iter}%6d] Metric: [$metric] value: ${metrics(metric)}")
+                }
+
+              (model, metrics)
+            }
+
+            val finalMetrics = perIterationMetrics.last._2
+
+            perModelMetrics += (regularizationWeight -> finalMetrics)
+            sendEvent(
+              PhotonOptimizationLogEvent(
+                regularizationWeight,
+                Some(optimizationStatesTracker),
+                Some(perIterationMetrics),
+                Some(finalMetrics)))
       }
     } else {
-      lambdaModelAndTrackers.foreach { case (regularizationWeight: Double, glm: GeneralizedLinearModel, optimizationStatesTrackerOpt: Option[OptimizationStatesTracker]) =>
-        logger.info(s"Model with lambda = $regularizationWeight:")
+      lambdaModelAndTrackers.foreach { case (lambda, model, tracker) =>
 
-        val finalMetrics = Evaluation.evaluate(glm, validationData)
+          logger.info(s"Model with lambda = $lambda:")
 
-        finalMetrics
-          .keys
-          .toSeq
-          .sorted
-          .foreach { metric =>
-            logger.info(f"Metric: [$metric] value: ${finalMetrics(metric)}")
+          val trackerOpt = if(params.enableOptimizationStateTracker) {
```

Comment:
It can still be `false`, however in that case the optimization states are still tracked, they're just ignored afterwards. That's why this parameter is marked as deprecated - if we're always tracking the optimization states, we may as well dump the results to the log.